domain/cloud/state/state.go

```diff
@@ -640,97 +640,107 @@ func updateAuthTypes(ctx context.Context, tx *sql.Tx, cloudUUID string, authType
 		authTypeIds[i] = id
 	}
 
-	authTypeIdsBinds, authTypeIdsAnyVals := database.SliceToPlaceholder(authTypeIds)
-
 	// Delete auth types no longer in the list.
-	deleteQuery := fmt.Sprintf(`
+	deleteQuery, err := sqlair.Prepare(`
 DELETE FROM  cloud_auth_type
-WHERE        cloud_uuid = ?
-AND          auth_type_id NOT IN (%s)
-`, authTypeIdsBinds)
+WHERE        cloud_uuid = $M.cloud_uuid
+AND          auth_type_id NOT IN ($AuthTypeIds[:])
+`, authTypeIds, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	args := append([]any{cloudUUID}, authTypeIdsAnyVals...)
-	if _, err := tx.ExecContext(ctx, deleteQuery, args...); err != nil {
+	if err := tx.Query(ctx, deleteQuery, authTypeIds, sqlair.M{"cloud_uuid": cloudUUID}).Run(); err != nil {
 		return errors.Trace(err)
 	}
 
-	insertQuery := `
+	insertStmt, err := sqlair.Prepare(`
 INSERT INTO cloud_auth_type (uuid, cloud_uuid, auth_type_id)
-VALUES (?, ?, ?)
+VALUES ($CloudAuthType.uuid, $CloudAuthType.cloud_uuid, $CloudAuthType.auth_type_id)
 ON CONFLICT(cloud_uuid, auth_type_id) DO NOTHING;
-	`
+	`, CloudAuthType{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
 	for _, a := range authTypeIds {
-		if _, err := tx.ExecContext(ctx, insertQuery, utils.MustNewUUID().String(), cloudUUID, a); err != nil {
+		if err := tx.Query(ctx, insertStmt, CloudAuthType{ID: utils.MustNewUUID().String(), CloudUUID: cloudUUID, AuthTypeID: a}).Run(); err != nil {
 			return errors.Trace(err)
 		}
 	}
 	return nil
 }
 
-func updateCACerts(ctx context.Context, tx *sql.Tx, cloudUUID string, certs []string) error {
+func updateCACerts(ctx context.Context, tx *sqlair.TX, cloudUUID string, certs []string) error {
 	// Delete any existing ca certs - we just delete them all rather
 	// than keeping existing ones as the cert values are long strings.
-	deleteQuery := `
+	deleteQuery, err := sqlair.Prepare(`
 DELETE FROM  cloud_ca_cert
-WHERE        cloud_uuid = ?
-`
+WHERE        cloud_uuid = $M.cloud_uuid
+`, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	insertQuery, err := sqlair.Prepare(`
+INSERT INTO cloud_ca_cert (uuid, cloud_uuid, ca_cert)
+VALUES ($CloudCACert.uuid, $CloudCACert.cloud_uuid, $CloudCACert.ca_cert)
+`, CloudCACert{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	if _, err := tx.ExecContext(ctx, deleteQuery, cloudUUID); err != nil {
+	if err := tx.Query(ctx, deleteQuery, sqlair.M{"cloud_uuid": cloudUUID}).Run(); err != nil {
 		return errors.Trace(err)
 	}
 
-	insertQuery := `
-INSERT INTO cloud_ca_cert (uuid, cloud_uuid, ca_cert)
-VALUES (?, ?, ?)
-`
 	for _, cert := range certs {
-
-		if _, err := tx.ExecContext(ctx, insertQuery, utils.MustNewUUID().String(), cloudUUID, cert); err != nil {
+		if err := tx.Query(ctx, insertQuery, CloudCACert{ID: utils.MustNewUUID().String(), CloudUUID: cloudUUID, CACert: cert}).Run(); err != nil {
```

Comment:
Same here: it would be less cluttered/messy to make the CloudCACert struct first and then pass that to the query instead of creating the struct inline.

Reply:
Refactored, also refactored other places where this was the case.